### Reduction operation: the sum of the numbers in the range [0, a)

In [1]:
import time

def reduc_operation(a):
    """Compute the sum of the numbers in the range [0, a)."""
    x = 0
    for i in range(a):
        x += i
    return x

# Secuencial

value = 1000000

initialTime = time.time()
suma = reduc_operation(value)
finalTime = time.time()

print("Time taken by reduction operation:", (finalTime - initialTime), "seconds")

# Utilizando las operaciones mágicas de ipython
%timeit -r 2 reduc_operation(value)

print(f"\n \t Computing the sum of numbers in the range [0, value): {suma}\n")

Time taken by reduction operation: 0.03932070732116699 seconds
35.3 ms ± 1.72 ms per loop (mean ± std. dev. of 2 runs, 10 loops each)

 	 Computing the sum of numbers in the range [0, value): 499999500000



## Ejercicio 3.2 Apartado A

In [2]:
import time

def sumacomponentesdelista(a):
    """Compute the sum of the numbers in the range [0, a)."""
    x = 0
    for i in a:
        x += i
    return x

    
listaasumar=list(range(1000000))
##Primero usamos el bucle for
initialtime = time.time()
suma= sumacomponentesdelista(listaasumar)
finaltime=time.time()
tiempotot=finaltime-initialtime

print ("El numero total sumado es de "+str(suma))
print("El tiempo de ejecucion de la suma de la lista por bucle for es de:" +" " + str(tiempotot)+ " " + "segundos")
%timeit -r 2 sumacomponentesdelista(listaasumar)
##Ahora probamos con la funcion sum
initialtime = time.time()
suma= sum(listaasumar)
finaltime=time.time()
tiempotot=finaltime-initialtime

print("El tiempo de ejecucion de la suma de la lista por funcion sum es de:" +" " + str(tiempotot)+ " " + "segundos")
%timeit -r 2 sum(listaasumar)

El numero total sumado es de 499999500000
El tiempo de ejecucion de la suma de la lista por bucle for es de: 0.027173995971679688 segundos
26 ms ± 70.6 μs per loop (mean ± std. dev. of 2 runs, 10 loops each)
El tiempo de ejecucion de la suma de la lista por funcion sum es de: 0.0060939788818359375 segundos
5.89 ms ± 40.8 μs per loop (mean ± std. dev. of 2 runs, 100 loops each)


## Ejercicio 3.2 Apartado B

In [3]:
import numpy as np #pongo el alias mas usado 

lista_np=np.array(listaasumar)

##Primero usamos el bucle for
initialtime = time.time()
suma= sumacomponentesdelista(lista_np)
finaltime=time.time()
tiempotot=finaltime-initialtime

print ("El numero total sumado es de "+str(suma))
print("El tiempo de ejecucion de la suma de la lista por bucle for es de:" +" " + str(tiempotot)+ " " + "segundos")
%timeit -r 2 sumacomponentesdelista(lista_np)
##Ahora probamos con la funcion sum
initialtime = time.time()
suma= np.sum(lista_np)
finaltime=time.time()
tiempotot=finaltime-initialtime

print("El tiempo de ejecucion de la suma de la lista de numpy por funcion sum es de:" +" " + str(tiempotot)+ " " + "segundos")
%timeit -r 2 sum(lista_np)

El numero total sumado es de 499999500000
El tiempo de ejecucion de la suma de la lista por bucle for es de: 0.05763411521911621 segundos
52.3 ms ± 23.4 μs per loop (mean ± std. dev. of 2 runs, 10 loops each)
El tiempo de ejecucion de la suma de la lista de numpy por funcion sum es de: 0.00037288665771484375 segundos
42.2 ms ± 203 ns per loop (mean ± std. dev. of 2 runs, 10 loops each)


## Ejercicio 3.2 Apartado C
Con los resultados obtenidos podemos comprobar que el codigo original donde hace una suma en un rango tarda unos 0.039 segundos; muy parecido a cuando hacemos un bucle for para la suma recorriendo tanto una lista o el array generado con numpy. Siendo el tiempo del bucle for en lista de 0.027 segundos y el del bucle for en el array de numpy de 0.057 segundos. Mientras tanto al usar la funcion sum de listas vemos que el tiempo se reduce a 0.006 segundos; pero podemos observar que la forma mas rapida es el uso de la funcion sum de numpy sobre el array obteniendo un tiempo de 0.0003.